src/cached_object_store.rs

```diff
@@ -125,39 +182,112 @@ impl CachedObjectStore {
-            Ok(None) => {}
-            Err(_) => {
-                // TODO: add a warning log
-            }
-        };
-
-        // it's strange that GetOptions did not derive Clone. maybe we could add a derive(Clone) to object_store.
+        opts: &GetOptions,
+    ) -> object_store::Result<(Option<GetResult>, ObjectMeta, Attributes)> {
+        let mut opts = opts.clone();
         if let Some(range) = &opts.range {
             opts.range = Some(self.align_get_range(range));
         }
-
         let get_result = self.object_store.get_opts(location, opts).await?;
         let result_meta = get_result.meta.clone();
-        let result_attrs = get_result.attributes.clone();
-        // swallow the error on saving to disk here (the disk might be already full), just fallback
-        // to the object store.
-        // TODO: add a warning log here.
-        self.save_result(get_result).await.ok();
-        Ok((result_meta, result_attrs))
+        let result_attributes = get_result.attributes.clone();
+
+        let mut ghost_guard = self.ghost.lock().await;
+        // Disk cache miss, increase ghost queue capacity
+        self.adjust_ghost_capacity(false, &mut ghost_guard);
+        if ghost_guard.remove(location).is_some() {
+            // swallow the error on saving to disk here (the disk might be already full), just fallback
+            // to the object store.
+            // TODO: add a warning log here and process `save_result` failure for ghost queue
+            self.save_result(get_result).await.ok();
+            Ok((None, result_meta, result_attributes))
+        } else {
+            // Object is accessed for the first time, we will not store it in the disk cache
+            let meta = get_result.meta.clone();
+            ghost_guard.put(location.clone(), meta);
+            Ok((Some(get_result), result_meta, result_attributes))
+        }
+    }
+
+    // Local disk cache hit, read data from local cache
+    async fn local_range_hit(
+        &self,
+        location: &Path,
+        range: Option<GetRange>,
+        meta: ObjectMeta,
+        attributes: Attributes,
+    ) -> object_store::Result<GetResult> {
+        let range = self.canonicalize_range(range, meta.size)?;
+        let parts = self.split_range_into_parts(range.clone());
+
+        // read parts, and concatenate them into a single stream. please note that some of these part may not be cached,
+        // we'll still fallback to the object store to get the missing parts.
+        let futures = parts
+            .into_iter()
+            .map(|(part_id, range_in_part)| self.read_part(location, part_id, range_in_part))
+            .collect::<Vec<_>>();
+        let result_stream = stream::iter(futures).then(|fut| fut).boxed();
+
+        Ok(GetResult {
+            meta,
+            range,
+            attributes,
+            payload: GetResultPayload::Stream(result_stream),
+        })
+    }
+
+    fn adjust_ghost_capacity(
```

Comment:
what do you think about splitting this into two different functions `increase_ghost_capacity` and `decrease_ghost_capacity`? the intention of these functions would be more clear than `true` and `false` from a caller point of view. Having both paths in the same function doesn't really save that many lines of code.

Reply:
No problem.